### PR TITLE
EOSC Node AAI

### PR DESCRIPTION
**Pull request by enolfc**
# Summary

Implement the support for the personal projects as the primary group of the user so this can be properly reported in accounting.

Creates a new `EOSCNodeAuthenticator` that will try to get the primary group of the user from the entitlements by matching a `personal_project_re` regular expression with default value:
```
r"^urn:geant:eosc-federation.eu:group:pp:Personal%20Project%20Name-(.*)$"
```
If the RE contains a group, the match will be used as the name for the primary group of the user. If it does not match, the primary group will be the first group of the user that matches the `allowed_groups`.

Sample configuration
```
c.JupyterHub.authenticator_class = "eoscauthenticator"
c.EOSCNodeAuthenticator.allow_all = True
c.EOSCNodeAuthenticator.claim_groups_key = "eduperson_entitlement"
c.EOSCNodeAuthenticator.personal_project = r"^urn:mace:egi.eu:group:fedcloud-([^#]*)#.*$"
```

This will match `urn:mace:egi.eu:group:fedcloud-users#sso.egi.eu` and return `users` as primary group

Without groups in the re:
```
c.JupyterHub.authenticator_class = "eoscauthenticator"
c.EOSCNodeAuthenticator.allow_all = True
c.EOSCNodeAuthenticator.claim_groups_key = "eduperson_entitlement"
c.EOSCNodeAuthenticator.personal_project = r"^urn:mace:egi.eu:group:fedcloud-.*$"
```
This will match `urn:mace:egi.eu:group:fedcloud-users#sso.egi.eu` and return `urn:mace:egi.eu:group:fedcloud-users#sso.egi.eu` as primary group

---

**Related issue :**

### CODE REVIEW

**Comment by enolfc** [comment]
We would benefit from some automated testing here as we are bit blind on whether this works or not :(

**Review by andrea-manzi — CHANGES_REQUESTED** [review]
please check comment

File: egi_notebooks_hub/egiauthenticator.py [hunk 1]
```diff
@@ -149,6 +150,8 @@ def _validate_scope(self, proposal):
         too long.
         """,
     )
+    # reasonable defaults for Check-in
+    manage_groups = True
```
[the remark this answers is not in the document]
Comment (enolfc) [reply egi_notebooks_hub/egiauthenticator.py:154]: We used to fill in the groups from the user on our code, see:

https://github.com/EGI-Federation/egi-notebooks-hub/pull/121/files#diff-3109bc70e06463bd429c59fd3bdc1a0314e7ee8e6183ae7c1556b19f6e71e4a0L207

But now this is part of the oauthenticator: 
https://github.com/jupyterhub/oauthenticator/blob/72f95e7ac406cd3dc57baf26a0400b0c62456975/oauthenticator/oauth2.py#L1126

[the remark this answers is not in the document]
Comment (andrea-manzi) [reply egi_notebooks_hub/egiauthenticator.py:154]: aaaah ok, clear

**Comment by valtri** [comment]
Works like a charm.

The default regular expression for personal group is OK? In testing AAI is working this: `^urn:geant:eosc-federation.eu:group:(pp-.*)$`.

And maybe we could tune `login_service` variable for EGI and EOSC authenticator? Something like:
```
    @default("login_service")
    def _login_service_default(self):
        return "EGI Check-in"
```
...but that is just a cosmetics and overrode by html customizations anyway.

**Comment by enolfc** [comment]
> Works like a charm.
> 
> The default regular expression for personal group is OK? In testing AAI is working this: `^urn:geant:eosc-federation.eu:group:(pp-.*)$`.
> 
> And maybe we could tune `login_service` variable for EGI and EOSC authenticator? Something like:
> 
> ```
>     @default("login_service")
>     def _login_service_default(self):
>         return "EGI Check-in"
> ```
> 
> ...but that is just a cosmetics and overrode by html customizations anyway.

I think we can just define login_service in the EOSC one to something like "EOSC AAI", no need to define a default

**Review by valtri — APPROVED** [review]
LGTM!

**Review by andrea-manzi — APPROVED** [review]
LGTM! and tested by @valtri

**Pull request merged**